Automattic/woocommerce-payments
# Add card declined rate limiter in card checkout

Fixes part of 998-gh-Automattic/woocommerce-payments-server

#### Changes proposed in this Pull Request

Add card declined rate limiter in standard card checkout

This PR addresses only credit cards in the standard checkout (non UPE). I discovered some difficulties when implementing it for UPE and after a short chat with @marcinbot decided that for now this is what we are targeting.

Some of those difficulties can be found here 1058-gh-Automattic/woocommerce-payments-server#issuecomment-896776610
 
The threshold for the cooldown to kick in is set right now in 5 failed transactions.
The cooldown period is set for 10 minutes where all the payments will be blocked from being processed.

These are magic numbers and I can change it to anything :)

The errors that this includes are 'card_declined', 'incorrect_cvc', 'expired_card', 'incorrect_number', which I believe are pretty much all that we are interested in for this case.

After the cooldown kicks in this is the error that appears when trying to submit another payment.

<img width="719" alt="image" src="https://user-images.githubusercontent.com/2612426/130074729-3d339206-bc44-4b57-b859-57edbef85125.png">

There is also a note in the order when the rate limiter blocks a payment so merchants can get visibility about it 

<img width="293" alt="image" src="https://user-images.githubusercontent.com/2612426/130075024-60018c31-083f-4f63-8d6c-741305f7b523.png">

#### Testing instructions

* Go to checkout
* Pay with card declined test numbers, such as 4000000000000002, and see the card declined error.
* If you don't modify the parameters in the code for testing, you'll need to make 5 payments.
* See message of the screenshot.
* Wait the 10 minutes
* See the card declined error again.

-------------------
